### DISCUSSION TITLE
MIMXRT1024 EVK Board QSPI Flash issues

### DISCUSSION THREAD

**Discussion by PxGreen101** [zephyrproject-rtos/zephyr#42610] [Q&A] [answered]
There is an On-Chip 32 Mbit QSPI Flash is used on the MIMXRT1024

According to the reference manual:
24.5 SIP 4M Flash Memory
This chip is packaged with a flash memory Winbond W25Q32JV. For more details, please visit http://www.winbond.com.

This device has a jedec-id = [ef 40 16]; 

However I am looking at the latest nxp_rt1024.dtsi file and it is listed as a IS25WP032 which is an altogether different part and has an altogether different jedec-id

&flexspi {
	status = "okay";
	reg = <0x402a8000 0x4000>, <0x60000000 DT_SIZE_M(4)>;
	is25wp032: is25wp032@0 {
		compatible = "nxp,imx-flexspi-nor";
		size = <DT_SIZE_M(32)>;
		label = "IS25WP032";
		reg = <0>;
		spi-max-frequency = <133000000>;
		status = "okay";
		jedec-id = [9d 70 17];
		erase-block-size = <DT_SIZE_K(4)>;
		write-block-size = <1>;
	};
};

Is this a mistake? 

I tried modifying this definition for flexspi in version 2.6 (version I am using). That version has the device listed as a IS25WP064 which is incorrect but has been changed to the above in 3.0. 

I am trying to get the zephyr/samples/drivers/flash_shell to run with this development kit but the system hard faults as soon as I run any flash command. Has anyone successfully run this particular example with the MIMXRT1024 development board? 

![image](https://user-images.githubusercontent.com/98962032/153076018-dd2853d7-6500-4238-b55a-d8500463784f.png)

**Comment by danieldegrasse** [comment]
Based off what you are seeing your issue may be that the default flash device name selected by the example is incorrect. Can you try issuing `flash set_device IS25WP032` before using other flash commands?

#### More details on your issue, if the above doesn't help:

It does appear that the device name in the RT1024 EVK DTS is incorrect. However, the key property in that DTS file is `compatible = "nxp,imx-flexspi-nor";`. Using a subset of the patches from PR #38209, you can enable the flash driver on Zephyr 2.6. I've attached the patch I used: [0001-boards-mimxrt1024-Enable-flexspi.txt](https://github.com/zephyrproject-rtos/zephyr/files/8044268/0001-boards-mimxrt1024-Enable-flexspi.txt)

You also need to explicitly set the flash device name using `flash set_device`, since the default flash selection that the implementation available in Zephyr 3.0 makes is incorrect. Here's a set of commands working using the patch I have attached:

```
*** Booting Zephyr OS build zephyr-v2.6.0  ***
**No flash controller found!**
Run set_device <name> to specify one before using other commands.


uart:~$ flash set_device IS25WP064

uart:~$ flash page_read 1023 300
00 01 02 03 | 04 05 06 07 |08 09 0a 0b | 0c 0d 0e 0f
10 11 12 13 | 14 15 16 17 |18 19 1a 1b | 1c 1d 1e 1f
20 21 22 23 | 24 25 26 27 |28 29 2a 2b | 2c 2d 2e 2f
30 31 32 33 | 34 35 36 37 |38 39 3a 3b | 3c 3d 3e 3f
40 41 42 43 | 44 45 46 47 |48 49 4a 4b | 4c 4d 4e 4f
50 51 52 53 | 54 55 56 57 |58 59 5a 5b | 5c 5d 5e 5f
60 61 62 63 | 64 65 66 67 |68 69 6a 6b | 6c 6d 6e 6f
70 7b | 7c 7d 7ef
80 81 82 83 | 84 85 86 87 |88 89 8a 8b | 8c 8d 8e 8f
90 91 97 |98 99 9a 9b | 9c 9d 9e 9f
a0 a1 a2 a3 | a4 a5 a6 a7 |a8 a9 aa ab | ac ad ae af
b0 b1 b2 b3 | b4 b5 b6 b7 |b8 b9 ba bb | bc bd be bf
c0 c1 c2 c3 | c4 c5 c6 c7 |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff
ff ff ff ff 

uart:~$ flash page_erase 1023 4096
Erasing page 1023 (start offset 0x3ff000, size 0x1000)

uart:~$ flash page_read 1023 300
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff  ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ffff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
fff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff |ff ff ff ff | ff ff ff ff
ff ff ff ff | ff ff ff ff
ff ff ff ff 
uart:~$ 
```

**Reply by PxGreen101** [reply, marked as answer]
Thanks for letting me know about the flash set_device command.  

I had already modified the nxp_rt1024.dtsi to the correct device: W25Q32JVWJ0. I tried the command  flash set_device W25Q32JVWJ0

&flexspi {
	 status = "okay";
         reg = <0x402a8000 0x4000>, <0x60000000 DT_SIZE_M(4)>;  
         w25q32jvwj0: w25q32jvwj@0 {
		compatible = "nxp,imx-flexspi-nor";
		size = <33554432>;  
		label = "W25Q32JVWJ0";
		reg = <0>;
		spi-max-frequency = <133000000>;
		status = "okay";
		jedec-id = [ef 40 16];        /* Winbond part */
		erase-block-size = <4096>;
		write-block-size = <1>;
	};
};


Flash read and write commands work correctly now. Thanks.
